### PR TITLE
TEST: add duration report to tests, speed up two outliers

### PR DESCRIPTION
**Pull request by mattip**
Add duration report. On my machine there are a two tests that each take more than 5 seconds to run (out of a total of 70 secs), see if it repeats on the travis CI runs.

### CODE REVIEW

**Comment by mattip** [comment]
The travis report for vanilla python 3.6 before any test changes confirms there are two tests that together take 34 out of 133 secs.

<details>

```
========================== slowest 10 test durations ===========================
21.94s call     builds/venv/lib/python3.6/site-packages/numpy/testing/tests/test_utils.py::TestAssertNoGcCycles::test_fails
13.58s call     builds/venv/lib/python3.6/site-packages/numpy/lib/tests/test_histograms.py::TestHistogramOptimBinNums::test_scott_vs_stone
2.02s call     builds/venv/lib/python3.6/site-packages/numpy/f2py/tests/test_compile_function.py::test_f2py_init_compile[]
2.01s call     builds/venv/lib/python3.6/site-packages/numpy/f2py/tests/test_compile_function.py::test_f2py_init_compile[--noopt --debug]
1.49s call     builds/venv/lib/python3.6/site-packages/numpy/core/tests/test_mem_overlap.py::TestUFunc::test_unary_gufunc_fuzz
1.43s call     builds/venv/lib/python3.6/site-packages/numpy/core/tests/test_multiarray.py::TestAlignment::test_various_alignments
1.37s call     builds/venv/lib/python3.6/site-packages/numpy/f2py/tests/test_compile_function.py::test_f2py_init_compile[extra_args0]
1.25s call     builds/venv/lib/python3.6/site-packages/numpy/core/tests/test_scalarmath.py::TestModulus::test_float_modulus_exact
1.12s call     builds/venv/lib/python3.6/site-packages/numpy/core/tests/test_mem_overlap.py::test_internal_overlap_fuzz
1.11s setup    builds/venv/lib/python3.6/site-packages/numpy/f2py/tests/test_array_from_pyobj.py::TestIntent::test_in_out
7981 passed, 53 skipped, 171 deselected, 13 xfailed, 3 xpassed in 133.70 seconds
```

**Comment by mattip** [comment]
The test run that previously was 133 seconds is now 99

**Comment by tylerjereddy** [comment]
There's already a test duration report on Shippable / ARMv8 CI that should also be checked here.

Not sure if the tradeoffs are a clear win--don't have a strong view on the tests changed here. I think Eric mentioned he wrote one of them when we were talking about the slower ARM test times.

**Comment by mattip** [comment]
the shippable CI run went from 10min25sec to 9min31secs, a significant reduction in resources.

File: numpy/lib/tests/test_histograms.py [hunk 1]
```diff
@@ -554,15 +554,11 @@ def nbins_ratio(seed, size):
...
 
         # the average difference between the two methods decreases as the dataset size increases.
-        assert_almost_equal(abs(np.mean(ll, axis=0) - 0.5),
-                            [0.1065248,
-                             0.0968844,
-                             0.0331818,
-                             0.0178057],
-                            decimal=3)
+        avg = abs(np.mean(ll, axis=0) - 0.5)
+        assert_almost_equal(avg, [0.15, 0.09, 0.08, 0.03], decimal=2)
```
Comment (seberg) [inline numpy/lib/tests/test_histograms.py]: Sounds fine to me, @madphysicist do you think the test loses all meaning like this? If it does, we could mark it as slow as well...

File: numpy/testing/tests/test_utils.py [hunk 2]
```diff
@@ -1510,13 +1510,17 @@ class ReferenceCycleInDel(object):
...
 
             def __init__(self):
                 self.cycle = self
 
             def __del__(self):
                 # break the current cycle so that `self` can be freed
                 self.cycle = None
+                ReferenceCycleInDel.count += 1
+                if ReferenceCycleInDel.count > 5:
+                    ReferenceCycleInDel.make_cycle = False
```
Comment (seberg) [inline numpy/testing/tests/test_utils.py]: I do not think this works. It is supposed to hit a 100 depth limit and then run into the `RuntimeError`, but as is, I think it runs into the AssertionError instead. Maybe that AssertionError path should make sure it is on py2 to make the test more meaningful.

We could reduce the 100 iteration limit maybe (it does seem rather high). But probably the test is just slow, and we can mark it as such?

**Review by seberg — COMMENTED** [review]
The new testing stuff looks nice, the first test is fine (but give @madphysicist a bit time to comment), I think the second needs to be just marked as slow probably (although we could tweak things in any case).

File: numpy/testing/tests/test_utils.py (hunk 2 above)
Comment (mattip) [reply numpy/testing/tests/test_utils.py]: marked as slow

File: numpy/lib/tests/test_histograms.py (hunk 1 above)
Comment (seberg) [reply numpy/lib/tests/test_histograms.py]: OK, lets just change this to the fast version. If Joseph or so think it gets meaningless then, we can revert maybe.

File: numpy/testing/tests/test_utils.py (hunk 2 above)
Comment (seberg) [reply numpy/testing/tests/test_utils.py]: Sounds good, could adapt the testing code to fail on Py3, but it is a different issue.

**Pull request merged**